### PR TITLE
Adding identity type traits and update definition of identity_t alias

### PR DESCRIPTION
**Pull request by dhollman**
Rational was produced below in https://github.com/kokkos/kokkos/pull/3339#issuecomment-689716083

This change allows to derive from `identity` which arguably can be more readable in some situations.

Note that we considered that `type_identity` might be a better name since it is standardized in C++20 but we decided not to change for now.

### CODE REVIEW

**Review by dalg24 — COMMENTED** [review]
Drop 3b16267 and provide a meaning full commit message

**Comment by dhollman** [comment]
I'm breaking up the pull requests into a series, but if I don't have them as a sequence I can't rebase later commits onto something reasonable

**Comment by crtrott** [comment]
identity_t is only used in atomic files which will go away so I think if you want to rename to something more standard practice that would be fine.,

**Comment by crtrott** [comment]
I think we can leave identity_t for now and replace it later with stuff from desul.

**Review by dalg24 — CHANGES_REQUESTED** [review]
Provide a rational for that change in the description of the PR.  (Yes I read that the commit message says "small change to identity_t to enable inheritance".)

It is not clear to me why would one want to inherit from `identity`? 
How is
```C++
struct Foo : identity<Foo> {};
```
better than
```C++
struct Bar { using identity = Bar; };
```
?
I would argue that you hardly save any keystroke and you increase the cognitive load on the reader who needs to know what's `identity`.
C++20 adds `std::type_identity` but at that point you write more.

**TLDR** Please elaborate what advantage you'd get from inheriting from `identity` because it is non-obvious to me.

**Comment by dhollman** [comment]
I really don't want to fight over this. I don't care, really; this is how I've always written things, but if you don't like this, you can make the changes necessary and push to my branches. I don't have a problem with that, but I also don't see what in your objection justifies the amount of time we've spent on this. I'm sorry this has become a big issue, and I'm sorry for what part I've had in inflaming that. That said, I'll do my best to write a short justification here.

In short: I agree that in isolation, `Bar` is more readable than `Foo`. But when you have a sequence of 30 partial specializations of something (as we often do), it's easier to see that they are all type traits at a glance (especially if the `type` spans multiple lines, which it often does) if you use the `Foo` form—see <https://github.com/dhollman/kokkos/blob/bf4e2f32316984eb582e32a70e46556b424b0e25/core/src/Kokkos_ExecPolicy.hpp#L910-L942>. If you still don't believe me that this is more readable, I'm just going to have to ask you to trust me as a dyslexic person that it's easier to see there are no differences here. Beyond this, though, if we wanted to change the way type traits worked throughout Kokkos (for instance, deleting the default constructor might not be a bad idea), we only have to change it in one place.

**Review by dalg24 — APPROVED** [review]
